### PR TITLE
[opentracer] Span context propagation

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
**Note: this PR is dependent on #473.**

This PR adds support for span context propagation through `inject` and `extract` methods which in turn use a propagator for a given format to encode span contexts into a carrier.

Changes in this PR:
- `inject` and `extract` functions to the tracer
- `HTTPPropagator` which handles the encoding for the `TEXT_MAP` and `HTTP_HEADERS` formats
- Unit tests for the above

Note that the `BINARY` format is not supported at this time.

### CODE REVIEW

File: ddtrace/opentracer/tracer.py [hunk 1]
```diff
@@ -85,6 +117,40 @@ def start_span(self, operation_name=None, child_of=None, references=None,
...
+    def inject(self, span_context, format, carrier):
+        """Injects a span context into a carrier.
+
+        :param span_context: span context to inject.
+
+        :param format: format to encode the span context with.
+
+        :param carrier: the carrier of the encoded span context.
+        """
+        if not isinstance(carrier, dict):
```
Comment (palazzem) [inline ddtrace/opentracer/tracer.py]: why we're expecting this to be a `dict`? it should work with any kind of object, assuming there is a proper `propagator` defined. No?

File: ddtrace/opentracer/tracer.py [hunk 2]
```diff
@@ -85,6 +117,40 @@ def start_span(self, operation_name=None, child_of=None, references=None,
...
+        propagator.inject(span_context, carrier)
+
+    def extract(self, format, carrier):
+        """Extracts a span context from a carrier.
+
+        :param format: format that the carrier is encoded with.
+
+        :param carrier: the carrier to extract from.
+        """
+        if not isinstance(carrier, dict):
```
Comment (palazzem) [inline ddtrace/opentracer/tracer.py]: Same here.

File: ddtrace/opentracer/tracer.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply ddtrace/opentracer/tracer.py]: True, the check should be performed in the propagator where the type is implied, the tracer shouldn't care about the type of the carrier. I'll move these checks to the propagator.

Comment (palazzem) [reply ddtrace/opentracer/tracer.py]: Perfect!

File: ddtrace/opentracer/propagation/http.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            raise InvalidCarrierException('propagator expects carrier to be a dict')
+
+        ddspan_ctx = self._dd_propagator.extract(carrier)
+
+        baggage = {}
+        for key in carrier:
+            if key.startswith(HTTP_BAGGAGE_PREFIX):
+                baggage[key[HTTP_BAGGAGE_PREFIX_LEN:]] = carrier[key]
+            elif key not in VALID_BAGGAGE_KEYS:
+                raise SpanContextCorruptedException('invalid key in span context')
```
Comment (palazzem) [inline ddtrace/opentracer/propagation/http.py]: Do we need that? I guess a `carrier` could be an HTTP header. So it has `HTTP_HEADER_TRACE_ID` and other things. Some of them are related to the baggage, but if users are propagating custom things such as `HTTP_LOG_DEBUG_ID`? We're raising an exception all the time making impossible the extraction.

I think we should extract the span context with our internal propagator and then retrieve the baggage. Any other header or field must simply be ignored. Am I missing a use case?

File: ddtrace/opentracer/span_context.py [hunk 4]
```diff
@@ -1,5 +1,10 @@
 from opentracing import SpanContext as OpenTracingSpanContext
 
+from ddtrace.context import Context
```
Comment (palazzem) [inline ddtrace/opentracer/span_context.py]: I guess this is a conflict not properly solved? I see 3 times the same import.

Comment (Kyle-Verhoog) [reply ddtrace/opentracer/span_context.py]: Must be. Haha.

File: ddtrace/opentracer/propagation/http.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddtrace/opentracer/propagation/http.py]: No you're right. I overlooked with this check that the carrier may contain keys that are user defined. I'll remove the check.

**Pull request merged**